Fix explain plan ALL_SEGMENTS_PRUNED_ON_SERVER node

somandal opened the pull request:
Today EXPLAIN PLAN sets the operator ID and parent ID of ALL_SEGMENTS_PRUNED_ON_SERVER as 2 and 1 respectively. This PR modifies ALL_SEGMENTS_PRUNED_ON_SERVER to use the operator ID and parent ID as 3 and 2 respectively instead. This is to bring consistency with the other explain plan nodes. The root is BROKER_REDUCE which has the ID 1 and parent ID 0, followed by the COMBINE node which has ID 2 and parent ID 1.

For scenarios where all segments were pruned across all servers, added special handling to keep the ALL_SEGMENTS_PRUNED_ON_SERVER operator ID and parent ID as 2 and 1 respectively. This is because for such scenarios no COMBINE node is present.

vvivekiyer commented:
@somandal  You probably also have to fix it here - https://github.com/apache/pinot/blob/master/pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java#L2483 ?

somandal commented:
> @somandal You probably also have to fix it here - https://github.com/apache/pinot/blob/master/pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java#L2483 ?

good catch, thanks! fixed it

vvivekiyer commented:
As discussed offline, we should probably have a dummy passthrough node in the "all segments pruned" case so that we don't show a disconnected tree (for example: https://github.com/apache/pinot/blob/6d2a886f28909c6ab050ad5f15386eca70372abb/pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java#L2483 here we'll not have any node with operatorId=2 and parentId=1.

somandal commented:
> As discussed offline, we should probably have a dummy passthrough node in the "all segments pruned" case so that we don't show a disconnected tree.
> 
> https://github.com/apache/pinot/blob/6d2a886f28909c6ab050ad5f15386eca70372abb/pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java#L2483
> 
> In the above example, we'll not have any node with operatorId=2 and parentId=1 unless we have the dummy node.

good catch. i've fixed this now. also updated the to handle selecting the correct COMBINE node depending on if we have any which aren't PASSTHROUGH or not.

vvivekiyer reviewed (approved):
LGTM

siddharthteotia commented on pinot-core/src/test/java/org/apache/pinot/queries/ExplainPlanQueriesTest.java:
```diff
@@ -1614,7 +1614,7 @@ public void testSelectColumnsVariationsOfAndOperatorsVerbose() {
     result4.add(new Object[]{"PLAN_START(numSegmentsForThisPlan:2)", ExplainPlanRows.PLAN_START_IDS,
         ExplainPlanRows.PLAN_START_IDS});
-    result4.add(new Object[]{"ALL_SEGMENTS_PRUNED_ON_SERVER", 2, 1});
+    result4.add(new Object[]{"ALL_SEGMENTS_PRUNED_ON_SERVER", 3, 2});
```
Can we add a test for the scenario where everything got pruned across all servers and thus it should be 2, 1 because COMBINE will be absent.

somandal replied:
@siddharthteotia this file already has tests for this scenario. e.g.

```
    // All segments are pruned
    String query10 = "EXPLAIN PLAN FOR SELECT count(*) FROM testTable WHERE invertedIndexCol3 = 'roadrunner' AND "
        + "noIndexCol1 = 100 LIMIT 100";
    List<Object[]> result10 = new ArrayList<>();
    result10.add(new Object[]{"BROKER_REDUCE(limit:100)", 1, 0});
    result10.add(new Object[]{"PLAN_START(numSegmentsForThisPlan:4)", ExplainPlanRows.PLAN_START_IDS,
        ExplainPlanRows.PLAN_START_IDS});
    result10.add(new Object[]{"ALL_SEGMENTS_PRUNED_ON_SERVER", 2, 1});
    check(query10, new ResultTable(DATA_SCHEMA, result10));
```
